drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1167 @@
+						  enum modem_cellular_event evt)
+{
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		if (modem_cellular_gpio_is_enabled(&config->power_gpio)) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_POWER_OFF);
+		} else {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		if (modem_cellular_is_registered(data)) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_CARRIER_ON);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		modem_cellular_start_timer(data, K_SECONDS(2));
+		modem_chat_script_run(&data->chat, &net_stat_chat_script);
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_register_state_leave(struct modem_cellular_data *data)
+{
+	modem_cellular_stop_timer(data);
+	modem_chat_release(&data->chat);
+	return 0;
+}
+
+static int modem_cellular_on_carrier_on_state_enter(struct modem_cellular_data *data)
+{
+	net_if_carrier_on(modem_ppp_get_iface(data->ppp));
+	modem_chat_attach(&data->chat, data->dlci1_pipe);
+	modem_chat_script_run(&data->chat, &net_stat_chat_script);
+	modem_cellular_start_timer(data, K_SECONDS(4));
+	return 0;
+}
+
+static void modem_cellular_carrier_on_event_handler(struct modem_cellular_data *data,
+						    enum modem_cellular_event evt)
+{
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_CARRIER_OFF);
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		if (modem_cellular_is_registered(data) == false) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_RUN_DIAL_SCRIPT);
+		}
+
+		break;
+
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		modem_chat_script_run(&data->chat, &net_stat_chat_script);
+		modem_cellular_start_timer(data, K_SECONDS(4));
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_carrier_on_state_leave(struct modem_cellular_data *data)
+{
+	modem_cellular_stop_timer(data);
+	modem_chat_script_abort(&data->chat);
+	modem_chat_release(&data->chat);
+	modem_ppp_release(data->ppp);
+	return 0;
+}
+
+static int modem_cellular_on_carrier_off_state_enter(struct modem_cellular_data *data)
+{
+	net_if_carrier_off(modem_ppp_get_iface(data->ppp));
+	modem_cellular_start_timer(data, K_SECONDS(1));
+	return 0;
+}
+
+static void modem_cellular_carrier_off_event_handler(struct modem_cellular_data *data,
+						 enum modem_cellular_event evt)
+{
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		if (modem_cellular_gpio_is_enabled(&config->power_gpio)) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_POWER_OFF);
+			break;
+		}
+
+		if (modem_cellular_gpio_is_enabled(&config->reset_gpio)) {
+			gpio_pin_set_dt(&config->reset_gpio, 1);
+		}
```

Comment:
This should be changed so modems with a reset only line take the same path. Just hold the modem in reset for power off.